### PR TITLE
Write caas application provisioner worker doc.go

### PR DESCRIPTION
**Pull request by jack-w-shaw**
Write a doc.go file for caas application provisioner worker to improve our developer documentation

### CODE REVIEW

File: worker/caasapplicationprovisioner/doc.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// Package caasapplicationprovisioner is a state reconcilliation
```
Comment (barrettj12) [inline worker/caasapplicationprovisioner/doc.go:4]: reconcilliation -> reconciliation (only one `l`)

Also, could we describe in simpler terms what a "state reconciliation engine" is? E.g. "the caasapplicationprovisioner is responsible for reconciling ..."

File: worker/caasapplicationprovisioner/doc.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+// engine for caas. We create a sub-worker for each application
+// that exists in the model. These sub-workers drive the kubernetes
+// provider to create, manage and destroy k8s resources.
+//
+// The worker also reflects the state of created application resources
+// back into the db, such as application/unit status, application/unit
```
Comment (barrettj12) [inline worker/caasapplicationprovisioner/doc.go:10]: I'm assuming this is pulling data from k8s and updating the Juju DB. If so, can we make this more explicit.

I'd also suggest using "database" instead of "db" - IMO the abbreviation is not necessary here.

The examples are nice.

File: worker/caasapplicationprovisioner/doc.go [hunk 3]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// Package caasapplicationprovisioner is a state reconcilliation
+// engine for caas. We create a sub-worker for each application
```
Comment (barrettj12) [inline worker/caasapplicationprovisioner/doc.go:5]: Can you elaborate on what "sub-worker" means. Are you saying that there is a single "primary" `caasapplicationprovisioner` worker for the model, which manages a bunch of "subordinate" `caasapplicationprovisioner` workers, one per application in the model?

File: worker/caasapplicationprovisioner/doc.go [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+// that exists in the model. These sub-workers drive the kubernetes
+// provider to create, manage and destroy k8s resources.
```
Comment (barrettj12) [inline worker/caasapplicationprovisioner/doc.go:7]: kubernetes -> Kubernetes
k8s -> K8s (or write in full as "Kubernetes")

**Review by barrettj12 — CHANGES_REQUESTED** [review]
Great start, I think we could make this even better though.

File: worker/caasapplicationprovisioner/doc.go (hunk 4 above)
Comment (tmihoc) [reply worker/caasapplicationprovisioner/doc.go:7]: I'd say, write in full, always (unless it's a chat in Mattermost).

Comment (tmihoc) [reply worker/caasapplicationprovisioner/doc.go:7]: "The model" sounds as if we're talking about a specific model. Maybe "For each model, we..." or "Given a model, ..."? Or find another way to connect it to the previous sentence. For example, right now I'm not sure what the connection between the state reconciliation mentioned in the previous sentence and the model mentioned here is.

File: worker/caasapplicationprovisioner/doc.go (hunk 2 above)
Comment (tmihoc) [reply worker/caasapplicationprovisioner/doc.go:10]: What does "reflects" mean? Can we make it more specific?

File: worker/caasapplicationprovisioner/doc.go (hunk 1 above)
Comment (tmihoc) [reply worker/caasapplicationprovisioner/doc.go:4]: Might help with my next comment about the model.

**Review by tmihoc — CHANGES_REQUESTED** [review]

**Review by barrettj12 — APPROVED** [review]
Awesome!

**Review by tmihoc — APPROVED** [review]
Good job.

File: worker/caasapplicationprovisioner/doc.go [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// Package caasapplicationprovisioner defines two types of worker:
+//   - provisioner: Watches a Kubernetes model, and start a new worker
+//     of the appWorker whenever an application is created.
+//   - appWorker: Drives the Kubernetes provider to create, manage,
+//     and destroy resources. Also writes the state of created
+//     application resources back into the database, such as application/unit
+//     status, application/unit IP addresses & ports, filesystem info, etc.
+package caasapplicationprovisioner
```
Comment (tmihoc) [inline worker/caasapplicationprovisioner/doc.go:11]: Fixed a typo. Also made some small rephrase suggestions.

> // Package caasapplicationprovisioner defines two types of worker:
//   - provisioner: Watches a Kubernetes model and starts a new worker
//     of the appWorker type whenever an application is created.
//   - appWorker: Drives the Kubernetes provider to create, manage,
//     and destroy resources. Also writes the state of created
//     application resources -- application/unit status, 
//     application/unit IP addresses & ports, filesystem info, etc. --
//     back into the database.
package caasapplicationprovisioner

**Review by tmihoc — CHANGES_REQUESTED** [review]

**Review by tmihoc — APPROVED** [review]
LGTM, thanks!
